cardinal/search.go

```diff
@@ -1,47 +1,119 @@
 package cardinal
 
 import (
-	"pkg.world.dev/world-engine/cardinal/ecs"
-	"pkg.world.dev/world-engine/cardinal/types/entity"
+	"github.com/rotisserie/eris"
+	"pkg.world.dev/world-engine/cardinal/filter"
+	"pkg.world.dev/world-engine/cardinal/gamestate"
+	iterators2 "pkg.world.dev/world-engine/cardinal/iterators"
+	"pkg.world.dev/world-engine/cardinal/types"
+	"pkg.world.dev/world-engine/cardinal/types/engine"
 )
 
-// Search allowed for the querying of entities within a World.
+type cache struct {
+	archetypes []types.ArchetypeID
+	seen       int
+}
+
+// Search represents a search for entities.
+// It is used to filter entities based on their components.
+// It receives arbitrary filters that are used to filter entities.
+// It contains a cache that is used to avoid re-evaluating the search.
+// So it is not recommended to create a new search every time you want
+// to filter entities with the same search.
 type Search struct {
-	impl *ecs.Search
+	archMatches map[string]*cache
+	filter      filter.ComponentFilter
+	namespace   string
+	reader      gamestate.Reader
 }
 
-// SearchCallBackFn represents a function that can operate on a single EntityID, and returns whether the next EntityID
-// should be processed.
-type SearchCallBackFn func(EntityID) bool
+// NewSearch creates a new search.
+// It receives arbitrary filters that are used to filter entities.
+func NewSearch(wCtx engine.Context, filter filter.ComponentFilter) *Search {
+	return &Search{
+		archMatches: make(map[string]*cache),
+		filter:      filter,
+		namespace:   wCtx.Namespace(),
+		reader:      wCtx.StoreReader(),
+	}
+}
 
-// Each executes the given callback function on every EntityID that matches this search. If any call to callback returns
-// falls, no more entities will be processed.
-func (q *Search) Each(wCtx WorldContext, callback SearchCallBackFn) error {
-	err := q.impl.Each(
-		wCtx.Engine(), func(eid entity.ID) bool {
-			return callback(eid)
-		},
-	)
-	if wCtx.Engine().IsReadOnly() || err == nil {
-		return err
+type CallbackFn func(types.EntityID) bool
+
+// Each iterates over all entities that match the search.
+// If you would like to stop the iteration, return false to the callback. To continue iterating, return true.
+func (s *Search) Each(callback CallbackFn) error {
+	result := s.evaluateSearch()
+	iter := iterators2.NewEntityIterator(0, s.reader, result)
+	for iter.HasNext() {
+		entities, err := iter.Next()
+		if err != nil {
+			return err
```

Comment:
Each needs an IsReadOnly check, as well as a logAndPanic wrapper.

If Each encounters an error in a System, cardinal needs to panic. If Each encounters an error in a Query, it should be returned to the user.

Reply:
fyi, the the original Each also doesn't have an `isReadOnly` check. I would prefer this to be a separate PR if you want this to be changed.

Reply:
> Each needs an IsReadOnly check

What do you mean by `isReadOnly` check? Search _is_ read only, so it shouldn't error on when the context is read only?

Reply:
We're using IsReadOnly to tell when Each is being called from a System and when it's being called from a Query. 

In Queries, all search errors (e.g. a transient DB error) will be returned to the user's custom Query code, and they can fail query, or make up data, or do whatever they want. 

In Systems, that same transient DB error should NOT be returned to the user's custom System code. If they make a decision to update state or not based on that error, they will have a non-deterministic system.

In summary: nil errors can always safely be returned to the caller. "Deterministic" errors can always safely be returned to the caller.

For all other errors:
IsReadOnly == true? return it to the caller
IsReadOnly == false? logAndPanic